# Preliminary Exploratory Data Analysis 2: Weather Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Preprocessing

### Load Data

In [17]:
stations = pd.read_csv('data/ghcnd-stations.csv')

In [26]:
stations[(stations['LOC_ID'] == 'CA') & (stations['NAME'].str.match('LOS ANGELES'))]

ID  LATITUDE  LONGITUDE   VAL1 LOC_ID  \
56601   US1CALA0030   34.1712  -118.4353  205.7     CA   
56621   US1CALA0064   34.0809  -118.2729  123.4     CA   
96356   USC00045111   34.0500  -118.2500  125.0     CA   
96357   USC00045112   34.0667  -118.2333   85.0     CA   
120337  USW00023174   33.9381  -118.3889   29.6     CA   
120927  USW00093134   34.0511  -118.2353   70.1     CA   

                               NAME TYPE     VAL2  
56601           LOS ANGELES 4.3 NNW  NaN      NaN  
56621            LOS ANGELES 2.6 NW  NaN      NaN  
96356          LOS ANGELES 6TH MAIN  NaN      NaN  
96357        LOS ANGELES TERMINAL A  NaN      NaN  
120337          LOS ANGELES INTL AP  NaN  72295.0  
120927  LOS ANGELES DWTN USC CAMPUS  NaN      NaN

Best Single Weather Station ID:

|ID	|LATITUDE	|LONGITUDE	|VAL1	|LOC_ID	|NAME	|TYPE	|VAL2
|----	|--------	|--------	|----	|-------	|------	|------	|----
|USW00093134	|34.0511	|-118.2353	|70.1	|CA	|LOS ANGELES DWTN USC CAMPUS      |    |		


In [100]:
weather = pd.read_csv('data/USW00093134.csv', engine='python')

In [101]:
def get_date(row):
    string = str(row['DATE'])
    year = string[:4]
    month = string[4:6]
    day = string[6:]
    date = year + '-' + month + '-' + day
    return date

weather['S_DATE'] = weather.apply(get_date, axis=1)
weather['DATE'] = pd.to_datetime(weather['S_DATE'])
weather =  weather[['DATE', 'ELEMENT', 'DATA_VALUE']]

In [105]:
weather =  weather[['DATE', 'ELEMENT', 'DATA_VALUE']]

In [106]:
weather

DATE ELEMENT  DATA_VALUE
0      1906-04-01    TMAX         161
1      1906-04-02    TMAX         183
2      1906-04-03    TMAX         189
3      1906-04-04    TMAX         178
4      1906-04-05    TMAX         189
...           ...     ...         ...
289735 2022-08-07    WSF5          63
289736 2022-08-08    WSF5          54
289737 2022-08-09    WSF5          63
289738 2022-08-10    WSF5          63
289739 2022-08-01    WT01           1

[289740 rows x 3 columns]

In [ ]:
weather['DATE'] = pd.to_datetime(weather['DATE'])

In [67]:
pivot = weather.pivot_table(index='DATE', columns='ELEMENT', values='DATA_VALUE')

In [ ]:
pivot[~pivot['PRCP'].isna()]

In [52]:
weather

ID        DATE ELEMENT  DATA_VALUE M-FLAG Q-FLAG S_FLAG  \
0       USW00093134  1906-04-01    TMAX         161    NaN    NaN      6   
1       USW00093134  1906-04-02    TMAX         183    NaN    NaN      6   
2       USW00093134  1906-04-03    TMAX         189    NaN    NaN      6   
3       USW00093134  1906-04-04    TMAX         178    NaN    NaN      6   
4       USW00093134  1906-04-05    TMAX         189    NaN    NaN      6   
...             ...         ...     ...         ...    ...    ...    ...   
289735  USW00093134  2022-08-07    WSF5          63    NaN    NaN      W   
289736  USW00093134  2022-08-08    WSF5          54    NaN    NaN      W   
289737  USW00093134  2022-08-09    WSF5          63    NaN    NaN      W   
289738  USW00093134  2022-08-10    WSF5          63    NaN    NaN      W   
289739  USW00093134  2022-08-01    WT01           1    NaN    NaN      W   

        OBS_TIME  
0         2400.0  
1         2400.0  
2         2400.0  
3         2400.0  
4         2400.0  
...          ...  
289735       NaN  
289736       NaN  
289737       NaN  
289738       NaN  
289739       NaN  

[289740 rows x 8 columns]

In [61]:
weather.groupby(['ID', 'DATE']).max()

C:\Users\chief\AppData\Local\Temp\ipykernel_25384\4160057779.py:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.max is deprecated. In a future version, a TypeError will be raised. Before calling .max, select only columns which should be valid for the function.
  weather.groupby(['ID', 'DATE']).max()


ELEMENT  DATA_VALUE S_FLAG  OBS_TIME
ID          DATE                                           
USW00093134 1906-04-01    TMIN         161      6    2400.0
            1906-04-02    TMIN         183      6    2400.0
            1906-04-03    TMIN         189      6    2400.0
            1906-04-04    TMIN         178      6    2400.0
            1906-04-05    WT16         189      6    2400.0
...                        ...         ...    ...       ...
            2022-08-08    WSF5         317      W    2400.0
            2022-08-09    WSF5         333      W    2400.0
            2022-08-10    WSF5         333      W    2400.0
            2022-08-11    TMIN         311      D    2400.0
            2022-08-12    TMIN         306      D    2400.0

[39859 rows x 4 columns]

In [59]:
weather

ID        DATE ELEMENT  DATA_VALUE M-FLAG Q-FLAG S_FLAG  \
0       USW00093134  1906-04-01    TMAX         161    NaN    NaN      6   
1       USW00093134  1906-04-02    TMAX         183    NaN    NaN      6   
2       USW00093134  1906-04-03    TMAX         189    NaN    NaN      6   
3       USW00093134  1906-04-04    TMAX         178    NaN    NaN      6   
4       USW00093134  1906-04-05    TMAX         189    NaN    NaN      6   
...             ...         ...     ...         ...    ...    ...    ...   
289735  USW00093134  2022-08-07    WSF5          63    NaN    NaN      W   
289736  USW00093134  2022-08-08    WSF5          54    NaN    NaN      W   
289737  USW00093134  2022-08-09    WSF5          63    NaN    NaN      W   
289738  USW00093134  2022-08-10    WSF5          63    NaN    NaN      W   
289739  USW00093134  2022-08-01    WT01           1    NaN    NaN      W   

        OBS_TIME  
0         2400.0  
1         2400.0  
2         2400.0  
3         2400.0  
4         2400.0  
...          ...  
289735       NaN  
289736       NaN  
289737       NaN  
289738       NaN  
289739       NaN  

[289740 rows x 8 columns]